File: scripts/fastq_splitter/fastq_splitter.pl
#!/usr/bin/perl -w
#
#  FASTQ Splitter  -  a script for partitioning a FASTQ file into pieces
#
#  Version 0.2.1 (Jul 31, 2018)
#
#  Copyright (c) 2018 Shujia Huang
#
use Getopt::Long;
use strict;
use File::Basename;


my $start_time = time;

my ($opt_n_parts,$outdir,$opt_version,$uncompress,$opt_help);
my $opt_check = 1;
GetOptions("n-parts=i"   => \$opt_n_parts,
           "outdir=s"    => \$outdir,
           "uncompress"  => \$uncompress,
           "check"       => \$opt_check,
           "version"     => \$opt_version,
           "help"        => \$opt_help) or die "Can't parse command line arguments.\n";

sub show_version {
    print q{FASTQ Splitter 0.2.1 Copyright (c) 2018 Shujia Huang}, "\n";
}

sub show_help {
    print q{Usage: fastq-splitter.pl [options] <fastq file> ...
Options:
    --n-parts <INT>      - Divide into <INT> parts
    --outdir <STR>       - Output diretory. 
    --uncompress         - Uncompress output files.
    --check              - Check FASTQ correctness(1=>check, 0=>not check). [1]
    --version            - Show version.
    --help               - Show help.
},"\n";
}

if ($opt_version) { show_version(); }
if ($opt_help) { show_help(); }
if ($opt_help or $opt_version) {
    # show help message
    exit;
}

if (!defined($outdir)) {
    die "Error: Splitting directory is not specified ('--outdir' option)\n\n";
}

if (!defined($opt_n_parts)) {
    if (!$opt_help and !$opt_version) {
        show_version();
        show_help();
    }
    print STDERR "Error: Divide number is not specified ('--n-parts' option)\n\n";
    exit;
}

die "File for splitting is not specified\n" if !@ARGV;

if (defined($opt_n_parts) and $opt_n_parts <= 0) {
    die "Non-positive number of parts\n";
}

my $n_parts = defined($opt_n_parts) ? $opt_n_parts: 0;

if ($n_parts <= 1) {
    print "WARING: Divide into $n_parts parts, means we don't have to divide anything.\n\n";
    exit(0);
}

# globle value record line number for locating error info.
my $line_count;

foreach my $infile (@ARGV) {
    $line_count = 0;
    print "Loading $infile";
    split_file($infile, $n_parts, $outdir);
}
 
my $end_time = time;
my $elapsed_time = $end_time - $start_time;
print "** Fastq splitter is all done, elapsed $elapsed_time second", (($elapsed_time==1)?'':'s'), " elapsed **\n";

############################################################################################
####################################### Sub function #######################################
sub split_file {

    my ($infile, $n_parts, $outdir) = @_;
    if (!-e $infile or !-f $infile) {
        print "[Error] Can't find file: \"$infile\"\n";
        return;
    }

    my ($fqname, $ext);
    if ($infile =~ m/\.fastq.gz$/) {
        $fqname = basename($infile, ".fastq.gz");
        $ext = "fastq";
    } elsif ($infile =~ m/\.fastq$/) {
        $fqname = basename($infile, ".fastq");
        $ext = "fastq";
    } elsif ($infile =~ m/\.fq.gz$/) {
        $fqname = basename($infile, ".fq.gz");
        $ext = "fq";
    } elsif ($infile =~ m/\.fq$/) {
        $fqname = basename($infile, ".fq");
        $ext = "fq";
    } else {
        die "Error: Input files are not fastq files(*.fq.gz/*.fq/*.fastq.gz/*.fastq)\n";
    }
    $ext .= ".gz" if !$uncompress;

    # Get fastq size and read number
    my ($total_read_num, $total_base_num) = get_file_size($infile);
    my $time = time;
    my $elapsed_time = $time - $start_time;
    print ": $total_read_num sequences, $total_base_num bp, elapsed $elapsed_time seconds elapsed\n";
    print " => dividing into $n_parts parts\n";

    # split fastq
    open(my $IN, ($infile =~ m/\.gz/) ? "gzip -dc $infile |": $infile) or die "Error: Can't open file $infile\n";

    my $num_len = length($n_parts);
    my $read_num_per_file = ($total_read_num % $n_parts == 0) ? $total_read_num / $n_parts : int($total_read_num / $n_parts) + 1;
    for (my $part = 1; $part <= $n_parts; ++$part) {

        my $part_file = sprintf("%s.%0*d.%s",$fqname, $num_len, $part, $ext);
        my $out_sub_file = join("/", $outdir, $part_file);

        print "$out_sub_file\n";

        my $written = 0;
        my ($name_line, $read_seq, $plus_line, $read_qual);
        
        # Compress file by pigz is much much much better than gzip
        open my $OUT, ($out_sub_file =~ m/\.gz$/) ? "| pigz > $out_sub_file": ">$out_sub_file" or die "Error: Cannot write to $out_sub_file\n";
        while(($written < $read_num_per_file) and (!eof $IN)){

            # my ($read_len, $read) = get_fastq_read($IN, $opt_check);
            chomp($name_line=<$IN>);
            chomp($read_seq=<$IN>);
            chomp($plus_line=<$IN>);
            chomp($read_qual=<$IN>);

            my $read = join("\n", ($name_line, $read_seq, $plus_line, $read_qual));
            print $OUT "$read\n";
            ++$written;
        }
        close $OUT;

    }
    close $IN;
}

sub get_file_size {

    my ($file) = @_;
    open(my $IN, ($file =~ m/\.gz/) ? "gzip -dc $file | ": $file) or die "Error: Can't open file $file\n";

    my ($name_line, $read_seq, $plus_line, $read_qual);
    my ($total_read_num, $total_base_num) = (0, 0);
    $line_count = 0;
    while (!eof $IN) {

        # my ($read_len, $read) = get_fastq_read($IN, 1);
        # $total_base_num += $read_len;

        chomp($name_line=<$IN>);
        chomp($read_seq=<$IN>);
        chomp($plus_line=<$IN>);
        chomp($read_qual=<$IN>);
        $line_count += 4;

        ++$total_read_num;
        $total_base_num += length($read_seq);
    }
    close $IN;

    return ($total_read_num, $total_base_num);
}

sub get_fastq_read {
    my ($IN, $check) = @_;
    my $truncated = "Error: Incomplete FASTQ entry at the end -> looks like truncated input!\n";

    if (eof $IN) { return (0, ''); }
    chomp(my $name_line=<$IN>);
    ++$line_count;
    if (substr($name_line, 0, 1) ne '@') {
        die "Error parsing line $line_count: FASTQ entry does not start with '\@':\n$name_line";
    }

    if (eof $IN) { die $truncated; }
    chomp(my $read_seq = <$IN>);
    ++$line_count;
    my $read_len = length($read_seq);

    if (eof $IN) { die $truncated; }
    chomp(my $plus_line = <$IN>);
    ++$line_count;
    if (substr($plus_line,0,1) ne '+') {
        die "Error parsing line $line_count: Expecting '+', found '$plus_line'";
    }

    if (eof $IN) { die $truncated; }
    chomp(my $read_qual = <$IN>);
    ++$line_count;
    my $qual_len = length($read_qual);
    if ($qual_len != $read_len){
        if (eof($IN)) {
            die $truncated;
        } elsif ($check) {
            die "Error: Misformatted FASTQ entry in input line $line_count: 
            quality length ($qual_len) differs from sequence length ($read_len):\n$read_seq\n$read_qual\n";
        }
    }

    my $read = join("\n", ($name_line, $read_seq, $plus_line, $read_qual));
    return ($read_len, $read);
}

File: scripts/fastq_splitter/fastq_splitter.py
"""
FASTQ Splitter  -  a script for partitioning a FASTQ file into pieces

Version 0.1.1 (Aug 16, 2018)
Copyright (c) 2018 Shujia Huang
"""
import argparse
import os
import sys

from datetime import datetime

import gzip

version_description = "FASTQ Splitter 0.1.1 Copyright (c) 2018 Shujia Huang"

LINE_COUNT = 0
START_TIME = datetime.now()


def split_file(infile, n_parts, outdir):

    if not os.path.exists(infile):
        sys.stderr.write("Error: Can't find file: %s\n" % infile)
        sys.exit(1)

    fqname, ext = '', ''
    if infile.endswith(".fastq.gz"):
        fqname = os.path.basename(infile).split(".fastq.gz")[0]
        ext = "fastq.gz"
    elif infile.endswith(".fq.gz"):
        fqname = os.path.basename(infile).split(".fq.gz")[0]
        ext = "fq.gz"
    elif infile.endswith(".fastq"):
        fqname = os.path.basename(infile).split(".fastq")[0]
        ext = "fastq"
    elif infile.endswith(".fq"):
        fqname = os.path.basename(infile).split(".fq")[0]
        ext = "fq"
    else:
        sys.stderr.write("Error: The input files are not fastq format(*.fq.gz/*.fq/*.fastq.gz/*.fastq)\n")

    total_read_num, total_base_num = get_file_size(infile)

    elapsed_time = datetime.now() - START_TIME    
    print "Loaded %s: %d sequences, %d bp, %d seconds elapsed" % (infile, total_read_num, total_base_num, elapsed_time.seconds)
    print "=> dividing into %d parts:" % n_parts

    read_num_per_file = total_read_num/n_parts if total_read_num % n_parts == 0 else int(total_read_num/n_parts)+1
    num_len = len(str(n_parts))
    with gzip.open(infile) if infile.endswith(".gz") else open(infile) as I:
        for part in range(1, n_parts+1):
            part_file = "%s.%0*d.%s" % (fqname, num_len, part, ext)
            out_sub_file = '/'.join([outdir, part_file])
            
            print out_sub_file

            written = 0
            with gzip.open(out_sub_file, "wb") if out_sub_file.endswith(".gz") else open(out_sub_file, "w") as OUT:
                is_done = False
                while not is_done and written < read_num_per_file:
                    written += 1
                    is_done, _, read = get_fastq_read(I)
                    OUT.write("%s\n" % read)


def get_file_size(infile):

    total_read_num, total_base_num = 0, 0
    with gzip.open(infile) if infile.endswith(".gz") else open(infile) as I:

        is_done = False
        while not is_done:
            is_done, read_len, read = get_fastq_read(I)
            if not is_done:
                total_read_num += 1
                total_base_num += read_len

    return total_read_num, total_base_num


def get_fastq_read(iter_fh):
    truncated = "Error: Incomplete FASTQ entry at the end -> " \
                "looks like truncated input!"

    global LINE_COUNT

    name_line = fetch_next(iter_fh).strip()
    is_done = False if name_line else True
    if is_done:
        return True, 0, ""

    LINE_COUNT += 1
    if name_line[0] != "@":
        sys.stderr.write("Error parsing line %d: FASTQ entry does not start "
                         "with '@':\n%s\n" % (LINE_COUNT, name_line))

    read_seq = fetch_next(iter_fh).strip()
    LINE_COUNT += 1
    is_done = False if read_seq else True

    plus_line = fetch_next(iter_fh).strip()
    LINE_COUNT += 1
    is_done = False if plus_line else True

    if plus_line[0] != "+":
        sys.stderr.write("Error parsing line %d: Expecting '+', "
                         "found '%s'" % (LINE_COUNT, plus_line))

    read_qual = fetch_next(iter_fh).strip()
    LINE_COUNT += 1
    is_done = False if read_qual else True
    if len(read_seq) != len(read_qual):
        sys.stderr.write("Error: Misformatted FASTQ entry in input line %d: "
                         "quality length (%d) differs from sequence length "
                         "(%d):\n%s\n%s\n" % (LINE_COUNT, len(read_seq),
                                              len(read_qual), read_seq,
                                              read_qual))

    read = "\n".join([name_line, read_seq, plus_line, read_qual])
    return is_done, len(read_seq), read


def fetch_next(iter_fh):

    if iter_fh == '': return ''

    try:
        line = iter_fh.next()
    except StopIteration:
        line = ''

    return line


if __name__ == "__main__":

    parser = argparse.ArgumentParser(description=version_description)
    parser.add_argument("-i", "--infastq", dest='fastqfile',
                        help="Input fastq(or .fastq.gz) file")
    parser.add_argument("-n", "--n-parts", dest='nparts', type=int,
                        help="Divid into <n> parts. required")
    parser.add_argument("-d", "--outdir", dest="outdir",
                        help="Output directory. required")

    args = parser.parse_args()

    if args.outdir is None:
        parser.error("Splitting directory is not specified ('--outdir' option)\n")

    if args.nparts is None:
        parser.error("Divide number is not specified ('--n-parts' option)\n")

    if args.nparts <= 0:
        parser.error("Non-positive number of parts.\n")

    if args.nparts <= 1:
        sys.stderr.write("WARNING: Divide into %d parts, means we don't have to divide anything.\n" % args.nparts)
        sys.exit(0)

    args.fastqfile = os.path.abspath(args.fastqfile)
    args.outdir = os.path.abspath(args.outdir)

    split_file(args.fastqfile, args.nparts, args.outdir)

    elapsed_time = datetime.now() -START_TIME 
    print "\n** Fastq splitter done, %d seconds elapsed **\n" % elapsed_time.seconds












